src/Psl/Filesystem/delete_directory.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Filesystem;
+
+use FilesystemIterator;
+use Psl;
+use Psl\Fun;
+use Psl\Internal;
+use Psl\Iter;
+use Psl\Str;
+
+use function rmdir;
+
+/**
+ * Delete the directory specified by $directory.
+ *
+ * @throws Exception\RuntimeException                   If unable to delete the directory.
+ * @throws Psl\Exception\InvariantViolationException    If the directory specified by
+ *                                                      $directory does not exist.
+ */
+function delete_directory(string $directory, bool $recursive = false): void
+{
+    Psl\invariant(is_directory($directory), '$directory does not exists.');
+
+    if ($recursive && !is_link($directory)) {
+        Iter\apply(
+            new FilesystemIterator(
+                $directory,
+                FilesystemIterator::CURRENT_AS_PATHNAME | FilesystemIterator::SKIP_DOTS,
+            ),
+            Fun\when(
+                static fn(string $node) => is_directory($node),
+                static fn(string $node) => delete_directory($node, true),
+                static fn(string $node) => delete_file($node),
+            )
+        );
```

Comment:
this is the first time for me using `Fun` 😛

Reply:
It's fun to use! ;)